### PR TITLE
feat(hyd): Trim wheels control

### PR DESCRIPTION
**Pull request by crocket63**
Fixes #[issue_no]

## Summary of Changes

This PR implements the trim wheel electric motorisation and manual control.
The FBW demand will send a command to the electrical system which will regulate motor speed to match the angle demand.

Manual control of the wheels can be done, and should cause correct manual override feedback to the FBW system.

Hydraulic part will be a simple placeholder and full simulation will be a separate PR

- [x] Trim wheel position and speed
- [x] Electric motor regulation
- [x] Manual override mechanism
- [x] Placeholder hydraulic motors 
- [x] Input event handling
- [x] Hydraulic flow
- [x] Electrical busses

## Additional context

Discord username (if different from GitHub):

## Testing instructions

- FBW should control the elevator trim as usual in nominal flight.
- Trim can be adjusted manually by mouse draging / ingame keybindings. Any manual action should show on simvar A32NX_THS_MANUAL_CONTROL =1 when trim is manually used
- If only green or only yellow hydraulics available, manual trim action should result in slower (half the speed) THS movement.
- If no green and no yellow hydraulics, THS surface can't move manually or automatically, visually or on ECAM fctl screen.
- If no green and no yellow hydraulics, Trim wheel can move slightly up or down bt manual action, and then hits a limit (roughly + 30° - 30° rotation available), with no action on THS actual angle.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by BlueberryKing — COMMENTED** [review]
I've only had a quick glance (I'll get to reviewing soon I promise!) but I think we should rename all occurrences of `ths` to `trimmable_horizontal_stabilizer` to follow conventions. At least for filenames and structs. Local variables are ok to be called `ths`.

**Comment by crocket63** [comment]
> I've only had a quick glance (I'll get to reviewing soon I promise!) but I think we should rename all occurrences of `ths` to `trimmable_horizontal_stabilizer` to follow conventions. At least for filenames and structs. Local variables are ok to be called `ths`.

Isn't this a bit long for files and struct?  I mean if you don't know what ths is might better not looking into those files anyway :D

File: src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 1]
```diff
@@ -0,0 +1,1066 @@
...
+
+    fn write(&self, writer: &mut SimulatorWriter) {
+        writer.write(&self.manual_override_id, self.manual_override_active);
+    }
+}
+
+pub struct TrimmableHorizontalStabilizerAssembly {
+    pitch_trim_actuator: PitchTrimActuator,
+    trim_wheel: TrimWheels,
+    ths_assembly: TrimmableHorizontalStabilizerHydraulics,
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:519]: I don't think this should be called `ths_assembly` as it is not actually of type `TrimmableHorizontalStabilizerAssembly`, which is what I thought of when I saw `ths_assembly`.

File: src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 2]
```diff
@@ -0,0 +1,1066 @@
...
+    const HYDRAULIC_MOTOR_SPEED_REGULATION_COEF_MAP: [f64; 7] = [-1., -1., -0.6, 0., 0.6, 1., 1.];
+
+    // Gain to convert hyd motor speed to ths deflection speed
+    const HYD_MOTOR_SPEED_TO_THS_DEFLECTION_SPEED_GAIN: f64 = 0.000085;
+
+    const MAX_DEFLECTION_FOR_FULL_OPEN_SPOOL_VALVE_DEGREES: f64 = 0.4;
+
+    pub fn new(context: &mut InitContext, min_deflection: Angle, deflection_range: Angle) -> Self {
+        Self {
+            deflection_id: context.get_identifier("HYD_FINAL_THS_DEFLECTION".to_owned()),
```
Comment (BlueberryKing) [inline src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:628]: This SimVar should be documented.

File: src/fbw/src/FlyByWireInterface.cpp [hunk 3]
```diff
@@ -444,8 +444,10 @@ void FlyByWireInterface::setupLocalVariables() {
   idAileronPositionLeft = make_unique<LocalVariable>("A32NX_AILERON_LEFT_DEFLECTION_DEMAND");
   idAileronPositionRight = make_unique<LocalVariable>("A32NX_AILERON_RIGHT_DEFLECTION_DEMAND");
 
-  idElevatorPosition = make_unique<LocalVariable>("A32NX_ELEVATOR_DEFLECTION_DEMAND");
+  idThs1MotorActive = make_unique<LocalVariable>("A32NX_THS_1_ACTIVE_MODE_COMMANDED");
+  idThs1MotorCommand = make_unique<LocalVariable>("A32NX_THS_1_COMMANDED_POSITION");
 
+  idElevatorPosition = make_unique<LocalVariable>("A32NX_ELEVATOR_DEFLECTION_DEMAND");
```
Comment (lukecologne) [inline src/fbw/src/FlyByWireInterface.cpp:450]: Why is the elevator deflection demand Lvar added again? Shouldn't this already be there from the hyd flight controls impact PR?

File: src/systems/a320_systems/src/hydraulic/mod.rs [hunk 4]
```diff
@@ -5447,6 +5491,86 @@ impl SimulationElement for A320GravityExtension {
...
+            motor2_active_id: context.get_identifier("THS_2_ACTIVE_MODE_COMMANDED".to_owned()),
+            motor3_active_id: context.get_identifier("THS_3_ACTIVE_MODE_COMMANDED".to_owned()),
+
+            motor1_position_id: context.get_identifier("THS_1_COMMANDED_POSITION".to_owned()),
+            motor2_position_id: context.get_identifier("THS_2_COMMANDED_POSITION".to_owned()),
+            motor3_position_id: context.get_identifier("THS_3_COMMANDED_POSITION".to_owned()),
+
+            manual_control_active_id: context
+                .get_identifier("THS_MANUAL_CONTROL_ACTIVE".to_owned()),
+            manual_control_speed_id: context.get_identifier("THS_MANUAL_CONTROL_SPEED".to_owned()),
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic/mod.rs:5525]: Oops!!!
THS_MANUAL_CONTROL_SPEED and THS_MANUAL_CONTROL_ACTIVE are local vars only in wasm though

File: src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs (hunk 2 above)
Comment (crocket63) [reply src/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:628]: This one is only local to wasm, which value is written to ELEVATOR TRIM POSITION

File: src/fbw/src/FlyByWireInterface.cpp (hunk 3 above)
Comment (crocket63) [reply src/fbw/src/FlyByWireInterface.cpp:450]: Was like "wtf" but it's just moved from one line to the other :)

**Review by BlueberryKing — APPROVED** [review]
Looks good now! However, I can only review the Rust code, not the C++ or the XML.

**Comment by crocket63** [comment]
> Looks good now! However, I can only review the Rust code, not the C++ or the XML.

Thanks, I think Luke had a look over the C++!

**Review by lukecologne — APPROVED** [review]
Yep, XML and C++ look good as well

**Comment by crocket63** [comment]
Thanks for the help : let's go to EXP :)

**Pull request merged**